connman/src/storage.c

```diff
@@ -794,11 +796,60 @@ static void keyfile_cleanup(void)
 	keyfile_hash = NULL;
 }
 
+static int validate_path(const char *pathname)
+{
+	char resolved_path[PATH_MAX] = { 0 };
+
+	errno = 0;
+
+	/*
+	 * realpath() expands all symbolic links and references. Returns
+	 * ENOENT if pathname is NULL.
+	 */
+	if (!realpath(pathname, resolved_path)) {
+		/* File does not exist when saving, return 0.*/
+		if (errno == ENOENT)
+			return 0;
+
+		DBG("realpath() error %d/%s", errno, strerror(errno));
+		return -errno;
+	}
+
+	switch (errno) {
```

Comment:
What is this about? AFAICT the errors that realpath() intentionally returns have already been handled above and this would cases where realpath() has returned success but errno has still changed -> just noise internal to realpath() that should not be of interest to anybody, or am I missing something?

Reply:
It is the GNU extension of `realpath()`:￼ ￼￼https://github.com/sailfishos/connman/pull/12#discussion_r711955362 as it apparently is not documented well enough it can set error with a non-NULL return in case there is a broken symlink. Or are we missing something?

Reply:
"If the call fails with either EACCES or ENOENT and resolved_path is not NULL..."

To me that reads as:

    if(!realpath(foo,bar) && (errno==EACCESS || errno==ENOENT)) gnu_stuff_happened;

and nothing points towards errno having any meaningful value when realpath() succeeds i.e. non-null value is returned

Reply:
iow I still think you should get rid of this "check errno after success" block (and possibly also that earlier errno=0 while at it).

Reply:
Yep got a bit confused here. Those have been changed now.